Src/Couchbase.Tests/Core/ClusterControllerTests.cs

```diff
@@ -178,6 +178,35 @@ public void When_Same_Bucket_Requested_Twice_In_Parallel_Only_One_Bootstrap_Is_D
             Assert.IsNotNull(bucket1);
             Assert.AreSame(bucket1, bucket2);
         }
+
+        [Test]
+        public async Task When_Disposing_Bucket_In_Parallel_Does_Not_Dispose_Referenced_Bucket()
+        {
+            var clusterController = new ClusterController(_clientConfig);
+            var cluster1 = new Cluster(_clientConfig, clusterController);
+            var cluster2 = new Cluster(_clientConfig, clusterController);
+
+            IBucket bucket1 = null;
+            IBucket bucket2 = null;
+
+            var t1 = new Thread(() =>
+            {
+                Thread.Sleep(100); // Give thread2 time to open the bucket
+                using (bucket1 = cluster1.OpenBucket("default", "")) { }
+            });
+
+            t1.Start();
+
+            using (bucket2 = cluster2.OpenBucket("default", ""))
+            {
+                Thread.Sleep(100); // Sleep while thread1 disposes the bucket
+                await bucket2.ExistsAsync("Key"); // Used to throw ObjectDisposedException
+            }
+
+            t1.Join();
+
+            Assert.AreSame(bucket1, bucket2);
+        }
```

Comment:
In the previous version of the code, this unit test will throw an ObjectDisposedException when one thread disposes the bucket another thread has been given.